### ISSUE TITLE
Unable to configure Agent as 'any'

### ISSUE THREAD

**Issue by clonmac** [ossec/ossec-hids#595]
I just installed version 2.9.0-beta03 on my server (upgraded from latest release on ossec.net). I then removed the extracted source folder and unzipped it again and then proceeded to compile the source for the Windows Agent. I am looking to use a development version of OSSEC so that I can utilize the EventChannel fixes that were put in place from issue #457. I installed the agent on a Windows 7 PC and configured an agent on the server (Ubuntu Server 14.04LTS).

The problem is that when configuring an agent on the server, the only way I can get the agent to connect successfully is if I specify the client's IP explicitly. If I use a CIDR address or 'any' the agent won't connect. I receive the following error in ossec.log on the server:

ossec-remoted(1213): WARN: Message from <agentIPhere> not allowed

On the client side it shows this in the log:

2015/04/20 13:08:50 ossec-agent: WARN: Process locked. Waiting for permission...
2015/04/20 13:09:01 ossec-agent(4101): WARN: Waiting for server reply (not started). Tried: '<serverIPhere>'.
2015/04/20 13:09:03 ossec-agent: INFO: Trying to connect to server (<serverIPhere>:1514).

If I revert the server and agent back to the original version (official latest release from ossec.net), everything will work fine and I the agent will connect when I use 'any'. There is no firewall blocking anything on the server and it is ossec-remoted that is refusing the connection.

I have also tried version 2.9.0-beta02 and also the latest master here on github and have the same problem. I am not sure if it is a problem with how I am compiling it or if there is a bug somewhere. Let me know if you need any more information or if I can help with testing. I am in a pre-deployment scenario so I don't mind blowing away the server and starting from scratch. I've done that numerous time so far. When the agent on the server is configured with the static IP, everything works great and the EventChannel logs come across formatted perfectly. Thanks.

**Comment by clonmac** [comment]
I've done some more testing and I am beginning to think there is a bug somewhere. I've compiled the latest version here on github updated the server with it. I then compiled the Windows Agent using 'make TARGET=winagent'. No errors occurred during compilation. If I try to configure the agent as 'any' on the server for the Windows client, it will be refused by the server. Taking those same installations and configuring the agent on the server with a static IP address, it will be accepted successfully.

Then, I downloaded the source from (http://www.ossec.net/?page_id=19) and compiled the Windows agent from there using the gen_win.sh script. After it finished all I did was simply copied over the ossec-agent-eventchannel.exe binary to the existing Windows client that wasn't connecting. I stopped the services, copied and renamed the binary as ossec-agent.exe and restarted the service (with client.keys configured as 'any') and it connected just fine. So the difference is clearly in the code somewhere.

I am not sure if there is a problem with the differences in how they're compiled or if there is a problem in the actual code. I noticed there are some differences in how the server accepts the hash values from the agent in /shared/hash_op.c, and /os_crypto/shared/keys.c. But, I don't think that the reason why the server rejects the agent is because of the code on the server since I never changed that in my testing. I haven't narrowed down yet where the code the Windows agent handles the hash that it passes to the server. I'm suspecting it is somewhere in there.

Finally, one last bit of information is that the issue is only with the Windows agent. Any Linux agent I install with the latest github code will connect without a problem when configured as 'any'.

**Comment by clonmac** [comment]
So there is definitely a bug introduced somewhere since the 2.8.1 release with allowing a Windows agent to connect to the server when configured as 'any'. The official 2.8.1 release works fine, but doesn't have some of the proper Windows EventChannel log formatting that was correcting in issue #457 . So I created a new fork, and made a new branch based off the 2.8.1 tag and implemented some of those EventChannel changes into that version (sorry if this isn't proper github etiquette, I am still new to github). So I now have a working version that allows the Windows agent to connect when configured as 'any' and also formats the EventChannel logs properly. This fork that is working for me can be found here in case anyone else is in a similar situation as me and would like to use EventChannel logging, but has DHCP Windows clients on their network.

https://github.com/clonmac/ossec-hids.git

That being said, the bug this issue open about still stands in the current development version of OSSEC and should be corrected before 2.9 is officially released.

**Comment by awiddersheim** [comment]
Thanks for the info. I'll try to take a look at this soon.

**Comment by clonmac** [comment]
I did some more testing with this and ran wireshark with the agent running on Windows. With version 2.8.1 it looks like it is appending the agent ID properly to each of the UDP packets. The beginning of the data section of the UDP packet it has 5 bytes that are constant on each packet that the agent sends to the server followed by the byte 0x3a (a colon) and then the encrypted message. But when I start the agent on windows using the 2.9 version, it only has 0x3a appended to the data section of each UDP packet sent to the server and so the server rejects the packet. I haven't figured it out yet, but for some reason the agent isn't detecting that it is a dynamic IP. I'm guessing that the if(!isSingleHost(keys->keyentries[id]->ip) && isAgent) check in the msgs.c file is failing and that's why.

**Comment by awiddersheim** [comment]
Interesting. I haven't had a whole lot of free time lately to look at this. Apologies. One thing you might want to try is using `git-bisect` and compiling the Windows agent to find which commit between 2.8.1 and 2.9 that broke things.

**Comment by clonmac** [comment]
Turns out it is the 'isAgent' part of the check that is failing not the IP portion. I took out the '&& isAgent' part of the above if statement in msgs.c and it is now appending the agent ID properly to each UDP packet and connecting to the server successfully.

**Comment by awiddersheim** [comment]
Sounds like something maybe broke with the redo of the build system. This is what sets `isAgent` in `headers/shared.h`:

```
#ifdef CLIENT
#define isAgent 1
#else
#define isAgent 0
#endif
```

**Comment by awiddersheim** [comment]
I'm wondernig if this will fix:

```
diff --git a/src/Makefile b/src/Makefile
index e37945d..3946fa2 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1213,19 +1213,19 @@ win32/ui/%.o: win32/ui/%.c
        ${OSSEC_CC} ${OSSEC_CFLAGS} -UOSSECHIDS -DARGV0=\"ossec-win32ui\" -c $^ -o $@
 
 win32/ossec-agent.exe: win32/icon.o win32/win_agent.o win32/win_service.o ${syscheck_o} ${rootcheck_o} $(filter-out client-agent/main.o, $(filter-out client-agent/agentd.o, $(filter-out client-
-       ${OSSEC_CCBIN} -DARGV0=\"ossec-agent\" -DOSSECHIDS ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
+       ${OSSEC_CCBIN} -DCLIENT -DARGV0=\"ossec-agent\" -DOSSECHIDS ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
 
 win32/ossec-agent-eventchannel.exe: win32/icon.o win32/win_agent.o win32/win_service.o ${syscheck_o} ${rootcheck_o} $(filter-out client-agent/main.o, $(filter-out client-agent/agentd.o, $(filte
-       ${OSSEC_CCBIN} -DARGV0=\"ossec-agent\" -DOSSECHIDS -DEVENTCHANNEL_SUPPORT ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
+       ${OSSEC_CCBIN} -DCLIENT -DARGV0=\"ossec-agent\" -DOSSECHIDS -DEVENTCHANNEL_SUPPORT ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
 
 win32/ossec-rootcheck.exe: win32/icon.o win32/win_service_rk.o ${rootcheck_rk_o} ${ossec_libs}
-       ${OSSEC_CCBIN} -DARGV0=\"ossec-rootcheck\" ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
+       ${OSSEC_CCBIN} -DCLIENT -DARGV0=\"ossec-rootcheck\" ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
 
 win32/manage_agents.exe: win32/win_service_rk.o ${addagent_o} ${ossec_libs} ${ZLIB_LIB}
-       ${OSSEC_CCBIN} -DARGV0=\"manage-agents\" -DMA ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
+       ${OSSEC_CCBIN} -DCLIENT -DARGV0=\"manage-agents\" -DMA ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
 
 win32/setup-windows.exe: win32/win_service_rk.o win32/setup-win.o win32/setup-shared.o ${ossec_libs}
-       ${OSSEC_CCBIN} -DARGV0=\"setup-windows\" ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
+       ${OSSEC_CCBIN} -DCLIENT -DARGV0=\"setup-windows\" ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
 
 win32/setup-syscheck.exe: win32/setup-syscheck.o win32/setup-shared.o ${ossec_libs}
        ${OSSEC_CCBIN} -DARGV0=\"setup-syscheck\" ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
```

**Comment by awiddersheim** [comment]
Were you able to test this patch?

**Comment by clonmac** [comment]
I just tried it. It doesn't look like adding the -DCLIENT option in the Makefile resolved it. Still fails at the same check and doesn't append the agent ID to the packet when they're sent.

**Comment by awiddersheim** [comment]
You recompiled and re-installed? I would have thought that would have fixed it.

**Comment by awiddersheim** [comment]
Maybe also adding `-DCLIENT` here as well. Either way I think its a issue that came out of the changes made to the build process and would require changes to the `Makefile`:

https://github.com/ossec/ossec-hids/blob/master/src/Makefile#L1204

**Comment by clonmac** [comment]
Just recompiled it again with the extra -DCLIENT on line 1204 and still got the same issue. I'm not familiar enough with the Makefile format to assist troubleshooting here.

**Comment by awiddersheim** [comment]
Understood. Can you try one more patch? If this doesn't work I'll have to find some time to really sit down and look at what is going on:

```
diff --git a/src/Makefile b/src/Makefile
index e37945d..ff23a99 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -316,6 +316,10 @@ ifeq (${TARGET},agent)
        OSSEC_CONF_SRC=../etc/ossec-agent.conf
 endif
 
+ifeq (${TARGET},winagent)
+       DEFINES+=-DCLIENT
+endif
+
 ifeq (${TARGET},local)
        DEFINES+=-DLOCAL
        OSSEC_CONTROL_SRC=./init/ossec-local.sh
@@ -1240,7 +1244,7 @@ win32/resource.o: win32/ui/win32ui.rc
        ${OSSEC_WINDRES} -i $< -o $@
 
 win32/os_win32ui.exe: win32/resource.o win32/win_service_rk.o ${win32_ui_o} addagent/b64.o ${ossec_libs}
-       ${OSSEC_CCBIN} -DCLIENT -DARGV0=\"ossec-win32ui\" ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
+       ${OSSEC_CCBIN} -DARGV0=\"ossec-win32ui\" ${OSSEC_CFLAGS} $^ ${OSSEC_LDFLAGS} -o $@
 
 
 ####################
```

**Comment by cgzones** [comment]
Hi,
or maybe can you try to add -DCLIENT in line 612.

**Comment by clonmac** [comment]
Looks like adding -DCLIENT to line 612 did the trick. Thanks for all the help to the both of you!

**Comment by awiddersheim** [comment]
Great. Thanks @cgzones.

**Comment by awiddersheim** [comment]
This was fixed by #617.
